google/cloud/bigtable/data/_metrics/handlers/_base.py

```diff
@@ -0,0 +1,28 @@
+# Copyright 2023 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from google.cloud.bigtable.data._metrics.data_model import CompletedOperationMetric
+
+
+class MetricsHandler:
```

Comment:
I think you need to eagerly publish attempt completion. Otherwise you will skew the client QPS from server QPS. ie its very useful to overlay client latency & qps graphs on top of serverside metrics to see deltas.

Reply:
Sure, I originally took eager attempt publishing out to simplify things (especially around types), and because I assumed eventually consistent was fine. But I added it back